rust/lance/src/session/caches.rs

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Caches for Lance datasets. They are organized in a hierarchical manner to
+//! avoid collisions.
+//!
+//!  GlobalMetadataCache
+//!     │
+//!     ├─► DSMetadataCache (prefixed by dataset URI)
+//!     │    │
+//!     └────┴──► FileMetadataCache (prefixed by file path)
+
+use std::{borrow::Cow, ops::Deref};
+
+use lance_core::{
+    cache::{CacheKey, LanceCache},
+    utils::{deletion::DeletionVector, mask::RowIdMask},
+};
+use lance_table::{
+    format::{DeletionFile, Manifest},
+    rowids::{RowIdIndex, RowIdSequence},
+};
+use object_store::path::Path;
+
+use crate::dataset::transaction::Transaction;
+
+/// A type-safe wrapper around a LanceCache that enforces namespaces for dataset metadata.
+pub struct GlobalMetadataCache(pub(super) LanceCache);
+
+impl GlobalMetadataCache {
+    pub fn for_dataset(&self, uri: &str) -> DSMetadataCache {
+        // Create a sub-cache for the dataset by adding the URI as a key prefix.
+        // This prevents collisions between different datasets.
+        DSMetadataCache(self.0.with_key_prefix(uri))
+    }
+
+    /// Create a file-specific metadata cache with the given prefix.
+    /// This is used by file readers and other components that need file-level caching.
+    pub(crate) fn file_metadata_cache(&self, path: &Path) -> LanceCache {
+        self.0.with_key_prefix(path.as_ref())
+    }
+}
+
+impl Clone for GlobalMetadataCache {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+/// A type-safe wrapper around a LanceCache that enforces namespaces and keys
+/// for dataset metadata.
+pub struct DSMetadataCache(pub(crate) LanceCache);
+
+impl Deref for DSMetadataCache {
+    type Target = LanceCache;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+// Cache key types for type-safe cache access
+#[derive(Debug)]
+pub struct ManifestKey<'a> {
+    pub version: u64,
+    pub e_tag: Option<&'a str>,
+}
+
+impl CacheKey for ManifestKey<'_> {
+    type ValueType = Manifest;
+
+    fn key(&self) -> Cow<'_, str> {
+        if let Some(e_tag) = self.e_tag {
+            Cow::Owned(format!("manifest/{}/{}", self.version, e_tag))
+        } else {
+            Cow::Owned(format!("manifest/{}", self.version))
+        }
+    }
+}
+
+#[derive(Debug)]
```

Comment:
Maybe just comment each of these keys and mention where they are used and why the various properties are there?